### PR TITLE
Add CircleCI support with CMake

### PR DESCRIPTION
**Pull request by danrbailey**
@Idclip - this adds Circle CI support to your CMake changes, I thought I'd open this as a PR into your fork rather than add these changes directly because there's quite a lot of them.

This supports building standalone ABI=6, ABI=5, ABI=4 as well as Houdini 17.5, 17.0, 16.5. It builds all of it's dependencies from scratch right now to try and best match the VFX Reference Platform, however there are a few remaining things to resolve:

* CMake 3.10.1 was the first release to support Boost 1.66 using FindBoost, so I think I need to update the version of CMake used in order to support the Boost version required in VFX Reference Platform 2019. I believe it picks up the default which is 3.7.3. Perhaps it's worth also using the earliest version of CMake as well for the other years, just to make sure that we're fully compatible.
* OpenEXR only introduced CMake support in 2.3.0 and I haven't got around to adding support to build it prior to that, so can't fully support VFX 2018/2017 yet.
* Houdini 17.5 ships with Imath/OpenEXR namespaced such as libImath_sidefx.so which I don't believe the CMake FindIlmBase modules can handle yet - I'm also not sure about the best way of supporting this?
* Docker images - I'd like to publish Docker images for VFX 2019/2018/2017 that has pre-installed all the non-VFX dependencies (Boost, TBB, etc), then re-use them here. It may even be possible to build out a latest top-of-tree build using Docker too.
* OpenEXR currently builds the whole library which includes some slow portions (the DWA compressor in particular). I'd like to push a change to OpenEXR to be able to optionally disable this when building as we don't support writing EXRs with this compression scheme in OpenVDB.

Also, I'd like to add a test job that builds everything at the latest top-of-tree. Ideally this is something we could run once a week or once a day or so to verify that any changes coming down the pipe don't cause issues for us.

### CODE REVIEW

**Comment by Idclip** [comment]
This looks good Dan, thanks! A few things in reply to your comments:

> Houdini 17.5 ships with Imath/OpenEXR namespaced such as libImath_sidefx.so which I don't believe the CMake FindIlmBase modules can handle yet - I'm also not sure about the best way of supporting this?

This is the same situation as boost no? We don't want to/can't use the namespaced packages so our own version of ilmbase/openexr now has to be provided. This is configured in  `OpenVDBHoudiniSetup` which only points ilmbase/openexr location to the houdini directory if HV <= 17.0

> OpenEXR only introduced CMake support in 2.3.0 ... VFX 2018/2017 yet.

Also need C++14 support in the cmake for this - and 2.3.0 is only for VFX 19 no?

**Comment by danrbailey** [comment]
Ah yes, you're right about boost. I thought for some reason SideFX were only shipping the boost libraries, but I realize the headers **are** included, just under "hboost" now. I missed that.

Why can't we use the namespaced packages? We wouldn't want to in a big studio, we'd want to build these libraries ourselves, but I don't see a reason why we should require someone to download boost and openexr if they already have it in Houdini. I'm not quite sure what the logic should be - pick up the non-namespaced ones as preference and then fall back to the Houdini ones if they're not available?

As for OpenEXR 2.3.0, yeah I'm using this for all reference platforms versions right now, because CMake isn't available for 2.2.0, but I'll have to get both working.

You're right about C++14, that's missing too.

**Comment by Idclip** [comment]
> Why can't we use the namespaced packages?

AFAIK because they all use Houdini namespacing - i.e. you cant just use hboost, you'd have to replace every instance of boost::method with hboost::method. The same goes for openexr. Example: https://www.sidefx.com/docs/hdk/_h_d_k__changes_16_5.html We'd have to add macros in OpenVDB to allow you to switch the namespace being used, so currently it's impossible to use sesis libs.

**Comment by danrbailey** [comment]
> > Why can't we use the namespaced packages?
> 
> AFAIK because they all use Houdini namespacing - i.e. you cant just use hboost, you'd have to replace every instance of boost::method with hboost::method. The same goes for openexr. Example: https://www.sidefx.com/docs/hdk/_h_d_k__changes_16_5.html We'd have to add macros in OpenVDB to allow you to switch the namespace being used, so currently it's impossible to use sesis libs.

Right, so for Boost, that's the case, the houdini library and SOPs have includes like this:

```
#if UT_VERSION_INT >= 0x10050000 // 16.5.0 or later
#include <hboost/algorithm/string/case_conv.hpp>
#include <hboost/algorithm/string/trim.hpp>
#else
#include <boost/algorithm/string/case_conv.hpp>
#include <boost/algorithm/string/trim.hpp>
#endif
```

However, the core library always uses `<boost/...` so you'd have to build your own (or make all the changes listed in that H16.5 doc).

I don't think that's the case for IlmBase/OpenEXR though as the header directory isn't called something like "hOpenEXR" so we should still be able to build against the version shipped with Houdini, it would just require picking up a library that is suffixed _sidefx.so (from H17.5+) which the CMake currently doesn't handle?

**Comment by Idclip** [comment]
Nah it's not just the directory naming, they've namespaced the headers and symbols too. I have this comment in the cmake when building the vdb deps:

```
# Set core deps. Note that the order here is important. If we're building against
# Houdini 17.5 we must include OpenEXR and IlmBase deps first to ensure the normal
# namespaced headers are used over the Houdini 17.5 ones (which will be imported
# through targets like tbb which are still deployed with Houdini)
```

If you look at toolkit/include/OpenEXR/IlmBaseConfig.h and OpenEXRConfig.h:

```
...
#define IMATH_INTERNAL_NAMESPACE Imath_2_2_sidefx
#define IEX_INTERNAL_NAMESPACE Iex_2_2_sidefx
#define ILMTHREAD_INTERNAL_NAMESPACE IlmThread_2_2_sidefx
...
```

and 

```
...
#define OPENEXR_IMF_INTERNAL_NAMESPACE Imf_2_2_sidefx
...
```

It's actually a bit annoying they haven't moved the headers out of include/OpenEXR because of this as it enforces the include order

**Comment by danrbailey** [comment]
> Nah it's not just the directory naming, they've namespaced the headers and symbols too. 

Actually, the headers haven't been namespaced from what I can tell and although the symbols _have_ been namespaced, they're still available without the "internal" name too. See this comment in ImathNamespace.h for example:

```
// To ensure source code compatibility, the IMATH_NAMESPACE defaults to Imath
// and then "using namespace IMATH_INTERNAL_NAMESPACE;" brings all of the
// declarations from the IMATH_INTERNAL_NAMESPACE into the IMATH_NAMESPACE.
// This means that client code can continue to use syntax like Imath::V3f,
// but at link time it will resolve to a mangled symbol based on the
// IMATH_INTERNAL_NAMESPACE.
```

So we can completely ignore the internal namespace and the only change needed is to be able to link against libImath_sidefx.so (from H17.5 onwards) rather than libImath.so.

**Comment by Idclip** [comment]
I tried to build against this a while back and it didn't immediately work out the box so I ignored it in favour of users providing their own for the current cmake change set. Have tried to build with these/got it working? I suggest this doesn't go into this batch of cmake changes if it involves any core vdb changes which aren't built system related. The linking issue shouldn't be too difficult to fix if we don't use openexrs cmake, we can just add custom changed to our own Find modules

**Comment by danrbailey** [comment]
> I tried to build against this a while back and it didn't immediately work out the box so I ignored it in favour of users providing their own for the current cmake change set. Have tried to build with these/got it working? I suggest this doesn't go into this batch of cmake changes if it involves any core vdb changes which aren't built system related. The linking issue shouldn't be too difficult to fix if we don't use openexrs cmake, we can just add custom changed to our own Find modules

Yeah, I think that's the change that's needed. I had a quick go at extending this:

```
      LIST ( APPEND CMAKE_FIND_LIBRARY_SUFFIXES
        "-${OpenEXR_VERSION_MAJOR}_${OpenEXR_VERSION_MINOR}.so"
        )
```

but it didn't work and I'm not sure this is quite the right mechanism to use anyway. I suspect that multiple calls to FIND_LIBRARY() with different library names would be better. Ultimately, we just need to decide the precedence for library names and search paths, which I suggest for standalone builds should be:

* libImath.so
* libImath_2_3.so

And for Houdini 17.5 builds should be:

* libImath.so
* libImath_2_3.so
* libImath_sidefx.so

Basically, what I'm saying is that I think there also needs to be a fallback library suffix hint (ie "_sidefx") set by OpenVDBHoudiniSetup.cmake for H17.5.

There's the related problem that I think you can end up with headers being pulled from user/system locations and libraries from Houdini locations or vice versa depending on version suffix or not (I can't remember which way round this gets processed). Perhaps a better logic could be to look for the header first, then use whereever that's found to derive the search paths for the library?

If you have ilmbase/openexr installed as user/system headers _and_ through Houdini, I'm not sure which should take precedence to be honest. Perhaps the former?

**Comment by Idclip** [comment]
Okay I think we're talking about two slightly different things. There two issues with ilmbase/openexr:

1) The version suffixed builds which error as they can't find the correct lib name. This should already be fixed in the came improvements branch, let me know if it doesn't work. The Find modules should support this lib naming.

2) The Houdini 17.5 build which, afaik, has two problems. The first is at compile time where I was getting namespace errors due to mismatching headers. 

> So we can completely ignore the internal namespace and the only change needed is to be able to link against libImath_sidefx.so (from H17.5 onwards) rather than libImath.so.

When I originally tried this a while back it wouldn't build due to the different namespaces, not because of issues with finding libImath_sidefx (though that's also a problem).

The second issue is during linking where the lib _sidefx can't be found. I propose this can be fixed by just adding _sidefx to CMAKE_FIND_LIBRARY_SUFFIXES outside the Find modules and updating the IlmBase/OpenEXR ones to append to CMAKE_FIND_LIBRARY_SUFFIXES rather than overwrite or add it internally.

> There's the related problem that I think you can end up with headers being pulled from user/system locations and libraries from Houdini locations or vice versa depending on version suffix or not (I can't remember which way round this gets processed). 

Here's the ilmbase ordered priority search paths:

```
  ${ILMBASE_LIBRARYDIR}
  ${_ILMBASE_ROOT_SEARCH_DIR}
  ${PC_IlmBase_LIBDIR}
  ${SYSTEM_LIBRARY_PATHS}
```

The only way this module will ever pick up a custom lib such as houdini's is by manually specifying `ILMBASE_LIBRARYDIR` or a `ILMBASE_ROOT` (<Package>_ROOT won't work for houdini). This is what `OpenVDBHoudiniSetup` does and it does it in such a way that it should be impossible for the includes and libs to pick up different locations as you suggest unless you're providing your own values for these (see below) OR it actually fails to find the lib and falls back to pkg-config or system deps. This would be a bug and should not happen after we fix the _sidefx issue.

> If you have ilmbase/openexr installed as user/system headers and through Houdini, I'm not sure which should take precedence to be honest. Perhaps the former?

The Houdini libs are never used unless you explicitly tell the cmake to with either `OPENVDB_BUILD_HOUDINI_SOPS` (fyi I'm going to deprecate this in favour of `OPENVDB_BUILD_HOUDINI_PLUGIN` but retain it for backwards compatibility). or `USE_HOUDINI`. In both cases the cmake uses `OpenVDBHoudiniSetup` to switch to a purely houdini dep build for its included libs EXCEPT for boost (H>=16.5), ilmbase and openexr (H>=17.5). The cmake is flexible enough that every module exposes <Package>_LIBRARYDIR and <Package>_INCLUDEDIR variables so not only could you build against system libs and choose to use tbb from houdini, but you can point to headers/libs installed in custom locations (basically not <prefix>/include <prefix>/lib).

**Comment by Idclip** [comment]
> The version suffixed builds which error as they can't find the correct lib name. This should already be fixed in the came improvements branch, let me know if it doesn't work. The Find modules should support this lib naming.

I've just tried this again and it now seems to build... I need to run some more tests to verify this

**Comment by danrbailey** [comment]
That all makes sense. Yes you're right that in a Houdini build, Houdini libs should always take precedence _unless_ you explicitly provide your own root / librarydir variables. So if you have Houdini and your own IlmBase/OpenEXR installed, it'll always pick up the Houdini one. 

> In both cases the cmake uses `OpenVDBHoudiniSetup` to switch to a purely houdini dep build for its included libs EXCEPT for boost (H>=16.5), ilmbase and openexr (H>=17.5). 

So that's the only quibble I have - the only exception should be Boost. IlmBase and OpenEXR for all versions of Houdini should **always** use the Houdini deps (just to be sure we're on the same page here).

> The second issue is during linking where the lib _sidefx can't be found. I propose this can be fixed by just adding _sidefx to CMAKE_FIND_LIBRARY_SUFFIXES outside the Find modules and updating the IlmBase/OpenEXR ones to append to CMAKE_FIND_LIBRARY_SUFFIXES rather than overwrite or add it internally.

That sounds like a good solution to this problem. 

As a somewhat unrelated note, the suffix used for OpenVDB is "_sesi" but for all other libraries, it's "_sidefx". I'm not sure if there's any reason to be able to build just the OpenVDB Houdini library or SOPs against the OpenVDB library Houdini ships with, but I think we should either explicitly support this or explicitly disable this (if that's not the case already). I think it's a common misconception that this is how you need to build the OpenVDB SOPs currently, so even with this clearly documented, we should still make the CMake error if we are not supporting this workflow to avoid confusion.

**Comment by Idclip** [comment]
> So that's the only quibble I have - the only exception should be Boost. IlmBase and OpenEXR for all versions of Houdini should always use the Houdini deps (just to be sure we're on the same page here).

Yep I agree, I thought that we were actually unable to use the exr namespaced libs in 17.5 but I'm currently unable to reproduce the issue I originally had so this seems fine - in which case I'll remove the 17.5 branching logic in `OpenVDBHoudiniSetup` such that it always points to the houdini deps if they haven't been specified.

> As a somewhat unrelated note, the suffix used for OpenVDB is "_sesi" but for all other libraries, it's "_sidefx". I'm not sure if there's any reason to be able to build just the OpenVDB Houdini library or SOPs against the OpenVDB library Houdini ships with, but I think we should either explicitly support this or explicitly disable this (if that's not the case already)

Hmmmm I mean technically you should be able to do something like:

```
OPENVDB_INCLUDEDIR=$HT/include
OPENVDB_LIBRARYDIR=$HDSO
LIST ( APPEND CMAKE_FIND_LIBRARY_SUFFIXES "_sesi.so"  )
FIND_PACKAGE ( OPENVDB REQUIRED )
```

And it should find it. I'll try and find some time to verify this. If it works, we could introduce an option for it but its getting a bit confusing, i.e. build_core=OFF, build_houdini=ON, use_houdini_vdb=??

**Comment by danrbailey** [comment]
> Hmmmm I mean technically you should be able to do something like:
> 
> ```
> OPENVDB_INCLUDEDIR=$HT/include
> OPENVDB_LIBRARYDIR=$HDSO
> LIST ( APPEND CMAKE_FIND_LIBRARY_SUFFIXES "_sesi.so"  )
> FIND_PACKAGE ( OPENVDB REQUIRED )
> ```
> And it should find it. I'll try and find some time to verify this. If it works, we could introduce an option for it but its getting a bit confusing, i.e. build_core=OFF, build_houdini=ON, use_houdini_vdb=??

I'm leaning towards explicitly not allowing it. If you check out the OpenVDB codebase at the exact same version as Houdini is using and just build the OpenVDB SOPs (plus any extra ones you might have written), sure you'll have no issues. However, as soon as there's a deviation in API because you build from a slightly different version than Houdini is shipping with, it's going to cause all kinds of issues that might be confusing to debug. I think we should explicitly disallow it and if you attempt to work around it by adding a libopenvdb.so symlink to libopenvdb_sesi.so (as an example of a hack an uninformed user might try), it should still catch this and explain why it's not recommended practice, forcing you to actually modify the CMake to switch off this check if you _really_ want to do this.

**Comment by danrbailey** [comment]
I just tested out your changes on the cmake_improvements branch and running into an issue where the function to retrieve the houdini default install directory doesn't exist in H16.5 or earlier:

```
CMake Error at openvdb_houdini/CMakeLists.txt:99 (houdini_get_default_install_dir):
  Unknown CMake command "houdini_get_default_install_dir".
```

**Comment by Idclip** [comment]
> I just tested out your changes on the cmake_improvements branch and running into an issue where the function to retrieve the houdini default install directory doesn't exist in H16.5 or earlier:

Thanks for catching this, I've fixed this on the cmake_stage branch along with a final few improvements. Performing a last few checks and will update the cmake_improvements branch with those. Last on my list is to see how much I can improve the docs by

**Comment by danrbailey** [comment]
I just checked your changes and it works fine now thanks.

One suggestion - I'd rather we leave Blosc enabled by default. Even though it's an optional dependency, OpenVDB relies so heavily on it, I think it could be confusing that you have to explicitly switch it on when building the core library as standalone.

I have added debug and GCC builds to Circle now, so it's almost matching Travis perfectly. The only thing it's missing is the ability to do header checks which I don't believe is supported in CMake? This is where you generate a .cpp for every header file which just #includes that one header and verify that it can be compiled.

I tried standing up an AppVeyor Windows build, but am running into this error:

```
CMake Warning (dev) at cmake/FindIlmBase.cmake:121 (IF):
  Policy CMP0057 is not set: Support new IN_LIST if() operator.  Run "cmake
  --help-policy CMP0057" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.
  IN_LIST will be interpreted as an operator when the policy is set to NEW.
  Since the policy is not set the OLD behavior will be used.
Call Stack (most recent call first):
  C:/Tools/vcpkg/scripts/buildsystems/vcpkg.cmake:247 (_find_package)
CMake Error at cmake/FindIlmBase.cmake:121 (IF):
  if given arguments:
    "NOT" "Half" "IN_LIST" "_ILMBASE_COMPONENT_LIST"
  Unknown arguments specified
```
I may have mis-understood this error, but I attempted to add a policy to the CMakeLists.txt but that doesn't seem to have worked.

Finally, I attempted to add a "latest" build running off the GitHub top-of-tree for every dependency and found two problems:
* Boost 1.70.0 beta is not yet fully supported in the latest CMake, I hard-coded "latest" to 1.69.0 and this resolved this one but it's not ideal. The association between Boost and CMake is a bit strange in my opinion.
* Using the latest Blosc produces a compile error in TestFile.cc that needs to be tracked down:

```
/root/project/openvdb/unittest/TestFile.cc:2555:17: error: no matching function for call to 'blosc_compcode_to_compname'
        if (0 > blosc_compcode_to_compname(compcode, &compname)) continue;
                ^~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/OpenEXR/../blosc.h:389:18: note: candidate function not viable: no known conversion from 'char **' to 'const char **' for 2nd argument
BLOSC_EXPORT int blosc_compcode_to_compname(int compcode, const char **compname);
```

I'm tempted to stand up a regular nightly (or even weekly?) job that builds OpenVDB against every released version of Blosc between 1.50 and the latest just so we can be 100% sure which versions are supported properly and to prevent any regressions in this area.

There's still a handful of todos in CMakeLists.txt such as this comment about malloc:

`# @todo scalable/concurrent malloc option(s)`

Are you planning to resolve these before merging?

**Comment by Idclip** [comment]
>  I'd rather we leave Blosc enabled by default. Even though it's an optional dependency, OpenVDB relies so heavily on it, I think it could be confusing that you have to explicitly switch it on when building the core library as standalone.

My opinion is that we keep it off. It's not available through apt-get and the version from Homebrew currently isn't tested and doesn't work with VDB (unit tests don't compile and fail even once fixed) so it might be a big ask for some users to have to manually download and build this. Ultimately I want to implement this whole "use if available" paradigm but it won't make it in to this PR. Perhaps something we can discuss at the TSC?

> The only thing it's missing is the ability to do header checks which I don't believe is supported in CMake?

Yeah I haven't done this - prioritising the documentation so will add this to my list of things to do.

> AppVeyor Windows build, but am running into this error

This is weird, is it using CMake >= 3.3? I'll try and have a look but thus far haven't had the chance to test on windows

>  The association between Boost and CMake is a bit strange in my opinion.

I agree, I'm not a fan either. I have some extra boost python/numpy fixes which I might try and get in to this PR too so if I do I can take a look at 1.7 as well.

> Using the latest Blosc produces a compile error in TestFile.cc that needs to be tracked down

Have a ticket for this here: https://jira.aswf.io/browse/OVDB-78 and is part of the reason I think Blosc should be off for the moment.

> todo's ... Are you planning to resolve these before merging?

Probably not, the only one I might try and resolve is the static/shared lib options but still prioritising the docs.

**Comment by danrbailey** [comment]
Thanks Nick, all sounds reasonable.

> My opinion is that we keep it off. It's not available through apt-get and the version from Homebrew currently isn't tested and doesn't work with VDB (unit tests don't compile and fail even once fixed) so it might be a big ask for some users to have to manually download and build this. Ultimately I want to implement this whole "use if available" paradigm but it won't make it in to this PR. Perhaps something we can discuss at the TSC?

I agree that it's not ideal, keeping it off is probably fine. In addition to the message that suggests using 1.5, perhaps it would be worth also adding a message that says that it's strongly recommended to use it if it's not enabled at all?

**Comment by Idclip** [comment]
Yeah sure that sounds goo, I'll do that.

One last thing, I still haven't enabled correct linking to _sidefx libs for 17.5. I've just tried this and am running into issues with vdb_render. All you need to add to enable this is:

```
# Versions of Houdini >= 17.5 have some namespaced libraries (IlmBase/OpenEXR).
# Add the required suffix as part of the cmake lib suffix searches

IF ( APPLE )
  LIST ( APPEND CMAKE_FIND_LIBRARY_SUFFIXES "_sidefx.so" )
ELSEIF ( UNIX )
  LIST ( APPEND CMAKE_FIND_LIBRARY_SUFFIXES "_sidefx.dylib" )
ENDIF ()
```

I'm going to continue looking into this before I commit this change though

**Comment by danrbailey** [comment]
> Yeah sure that sounds goo, I'll do that.
> 
> One last thing, I still haven't enabled correct linking to _sidefx libs for 17.5. I've just tried this and am running into issues with vdb_render. All you need to add to enable this is:
> 
> ```
> # Versions of Houdini >= 17.5 have some namespaced libraries (IlmBase/OpenEXR).
> # Add the required suffix as part of the cmake lib suffix searches
> 
> IF ( APPLE )
>   LIST ( APPEND CMAKE_FIND_LIBRARY_SUFFIXES "_sidefx.so" )
> ELSEIF ( UNIX )
>   LIST ( APPEND CMAKE_FIND_LIBRARY_SUFFIXES "_sidefx.dylib" )
> ENDIF ()
> ```
> I'm going to continue looking into this before I commit this change though

I'm currently not building the Python module, binaries or unit tests for Houdini, only for the standalone builds. This seems to work switching off installing IlmBase/OpenEXR, so I think it only needs libHalf.so in fact. I _can_ switch it on again, but not sure if that's worth doing or not as it's only really testing the build system?

**Comment by Idclip** [comment]
Yeah it does only need half, but what if you want to build the binaries against Houdini? This used to be "supported" in the sense that it worked. I now get this:

```
CMakeFiles/vdb_render.dir/openvdb_render.cc.o: In function `(anonymous namespace)::saveEXR(std::string const&, openvdb::v6_0abi5::tools::Film const&, (anonymous namespace)::RenderOpts const&)':
openvdb_render.cc:(.text+0x19a8): undefined reference to `Imf_2_2_sidefx::setGlobalThreadCount(int)'
openvdb_render.cc:(.text+0x19eb): undefined reference to `Imf_2_2_sidefx::Header::Header(int, int, float, Imath_2_2_sidefx::Vec2<float> const&, float, Imf_2_2_sidefx::LineOrder, Imf_2_2_sidefx::Compression)'
openvdb_render.cc:(.text+0x1a13): undefined reference to `Imf_2_2_sidefx::Header::compression()'
openvdb_render.cc:(.text+0x1a37): undefined reference to `Imf_2_2_sidefx::Channel::Channel(Imf_2_2_sidefx::PixelType, int, int, bool)'
openvdb_render.cc:(.text+0x1a43): undefined reference to `Imf_2_2_sidefx::Header::channels()'
openvdb_render.cc:(.text+0x1a57): undefined reference to `Imf_2_2_sidefx::ChannelList::insert(char const*, Imf_2_2_sidefx::Channel const&)'
openvdb_render.cc:(.text+0x1a75): undefined reference to `Imf_2_2_sidefx::Channel::Channel(Imf_2_2_sidefx::PixelType, int, int, bool)'
openvdb_render.cc:(.text+0x1a81): undefined reference to `Imf_2_2_sidefx::Header::channels()'
openvdb_render.cc:(.text+0x1a95): undefined reference to `Imf_2_2_sidefx::ChannelList::insert(char const*, Imf_2_2_sidefx::Channel const&)'
openvdb_render.cc:(.text+0x1ab3): undefined reference to `Imf_2_2_sidefx::Channel::Channel(Imf_2_2_sidefx::PixelType, int, int, bool)'
openvdb_render.cc:(.text+0x1abf): undefined reference to `Imf_2_2_sidefx::Header::channels()'
openvdb_render.cc:(.text+0x1ad3): undefined reference to `Imf_2_2_sidefx::ChannelList::insert(char const*, Imf_2_2_sidefx::Channel const&)'
openvdb_render.cc:(.text+0x1af1): undefined reference to `Imf_2_2_sidefx::Channel::Channel(Imf_2_2_sidefx::PixelType, int, int, bool)'
openvdb_render.cc:(.text+0x1afd): undefined reference to `Imf_2_2_sidefx::Header::channels()'
openvdb_render.cc:(.text+0x1b11): undefined reference to `Imf_2_2_sidefx::ChannelList::insert(char const*, Imf_2_2_sidefx::Channel const&)'
openvdb_render.cc:(.text+0x1b8d): undefined reference to `Imf_2_2_sidefx::Slice::Slice(Imf_2_2_sidefx::PixelType, char*, unsigned long, unsigned long, int, int, double, bool, bool)'
openvdb_render.cc:(.text+0x1ba8): undefined reference to `Imf_2_2_sidefx::FrameBuffer::insert(char const*, Imf_2_2_sidefx::Slice const&)'
openvdb_render.cc:(.text+0x1be0): undefined reference to `Imf_2_2_sidefx::Slice::Slice(Imf_2_2_sidefx::PixelType, char*, unsigned long, unsigned long, int, int, double, bool, bool)'
openvdb_render.cc:(.text+0x1bf8): undefined reference to `Imf_2_2_sidefx::FrameBuffer::insert(char const*, Imf_2_2_sidefx::Slice const&)'
openvdb_render.cc:(.text+0x1c30): undefined reference to `Imf_2_2_sidefx::Slice::Slice(Imf_2_2_sidefx::PixelType, char*, unsigned long, unsigned long, int, int, double, bool, bool)'
openvdb_render.cc:(.text+0x1c4b): undefined reference to `Imf_2_2_sidefx::FrameBuffer::insert(char const*, Imf_2_2_sidefx::Slice const&)'
openvdb_render.cc:(.text+0x1c83): undefined reference to `Imf_2_2_sidefx::Slice::Slice(Imf_2_2_sidefx::PixelType, char*, unsigned long, unsigned long, int, int, double, bool, bool)'
openvdb_render.cc:(.text+0x1c9b): undefined reference to `Imf_2_2_sidefx::FrameBuffer::insert(char const*, Imf_2_2_sidefx::Slice const&)'
openvdb_render.cc:(.text+0x1ca0): undefined reference to `Imf_2_2_sidefx::globalThreadCount()'
openvdb_render.cc:(.text+0x1cbc): undefined reference to `Imf_2_2_sidefx::OutputFile::OutputFile(char const*, Imf_2_2_sidefx::Header const&, int)'
openvdb_render.cc:(.text+0x1ccb): undefined reference to `Imf_2_2_sidefx::OutputFile::setFrameBuffer(Imf_2_2_sidefx::FrameBuffer const&)'
openvdb_render.cc:(.text+0x1cdb): undefined reference to `Imf_2_2_sidefx::OutputFile::writePixels(int)'
openvdb_render.cc:(.text+0x1dfd): undefined reference to `Imf_2_2_sidefx::OutputFile::~OutputFile()'
openvdb_render.cc:(.text+0x1e18): undefined reference to `Imf_2_2_sidefx::Header::~Header()'
openvdb_render.cc:(.text+0x1e6d): undefined reference to `Imf_2_2_sidefx::Header::compression()'
openvdb_render.cc:(.text+0x1e84): undefined reference to `Imf_2_2_sidefx::Header::compression()'
openvdb_render.cc:(.text+0x1fd3): undefined reference to `Imf_2_2_sidefx::Header::~Header()'
openvdb_render.cc:(.text+0x2099): undefined reference to `Imf_2_2_sidefx::OutputFile::~OutputFile()'
```

**Comment by Idclip** [comment]
I think it's just a problem with the CMake though, not the Houdini libs

**Comment by Idclip** [comment]
Okay I think this was just me being stupid, the snippet I sent you had the suffixes the wrong way around... All seems to work, situation normal....

**Comment by danrbailey** [comment]
> I think it's just a problem with the CMake though, not the Houdini libs

From the looks of things it needs the version in the suffix as well as "_sidefx":

```
    LIST ( APPEND CMAKE_FIND_LIBRARY_SUFFIXES
      "-${IlmBase_VERSION_MAJOR}_${IlmBase_VERSION_MINOR}.so"
      "-${IlmBase_VERSION_MAJOR}_${IlmBase_VERSION_MINOR}_sidefx.so"
      )
```

I suggest we just add the building of the binaries and Python module to the Clang Houdini 17.5 job, I'll do that now. I have a GCC H17.5 job, but I've had to throttle the threads to 1(!) to avoid it running out of memory so it takes 58 mins in comparison to the Clang build which takes just 24 mins.

**Comment by Idclip** [comment]
Pushed up this trivial change to the `cmake_stage` branch, will test on mac and travis too then commit

**Comment by Idclip** [comment]
So I'm still running into issues on Mac with Houdini 17.5.173, the same symbol errors as above. If I run `strings libIlmImf_sidefx.so | grep compression` on linux I get:

```
_ZN14Imf_2_2_sidefx6Header11compressionEv
_ZNK14Imf_2_2_sidefx6Header11compressionEv
_ZNK14Imf_2_2_sidefx14RgbaOutputFile11compressionEv
_ZNK14Imf_2_2_sidefx13RgbaInputFile11compressionEv
_ZNK14Imf_2_2_sidefx19TiledRgbaOutputFile11compressionEv
_ZNK14Imf_2_2_sidefx18TiledRgbaInputFile11compressionEv
_ZNK14Imf_2_2_sidefx14AcesOutputFile11compressionEv
_ZNK14Imf_2_2_sidefx13AcesInputFile11compressionEv
compression
Unknown compression type in image header.
" failed. The files use different compression methods.
Unhandled compression scheme case
Data compression (zlib) failed.
Data decompression (zlib) failed.
Invalid compression type for ACES file.
Unknown compression method in chunk offset reconstruction
_ZNK14Imf_2_2_sidefx13AcesInputFile11compressionEv
_ZNK14Imf_2_2_sidefx6Header11compressionEv
_ZNK14Imf_2_2_sidefx13RgbaInputFile11compressionEv
_ZNK14Imf_2_2_sidefx14AcesOutputFile11compressionEv
_ZNK14Imf_2_2_sidefx14RgbaOutputFile11compressionEv
_ZNK14Imf_2_2_sidefx18TiledRgbaInputFile11compressionEv
_ZN14Imf_2_2_sidefx6Header11compressionEv
_ZNK14Imf_2_2_sidefx19TiledRgbaOutputFile11compressionEv
```

on mac, `strings libIlmImf_sidefx.dylib | grep compression`, I get:

```
compression
Unknown compression type in image header.
" failed. The files use different compression methods.
Unhandled compression scheme case
Data compression (zlib) failed.
Data decompression (zlib) failed.
Invalid compression type for ACES file.
Unknown compression method in chunk offset reconstruction
```

**Comment by danrbailey** [comment]
I'll try a Mac build myself. I should point out that those symbols do exist:

If I do (note the extra hyphen) `strings - libIlmImf_sidefx.dylib | grep compression`:

```
compression
Unknown compression type in image header.
" failed. The files use different compression methods.
Data compression (zlib) failed.
Data decompression (zlib) failed.
Unhandled compression scheme case
Invalid compression type for ACES file.
Unknown compression method in chunk offset reconstruction
G1compressionEv
compressionEv
qcompressionEv
compressionEv
compressionEv
compressionEv
1compressionEv
1compressionEv
__ZN14Imf_2_2_sidefx6Header11compressionEv
__ZNK14Imf_2_2_sidefx13AcesInputFile11compressionEv
__ZNK14Imf_2_2_sidefx13RgbaInputFile11compressionEv
__ZNK14Imf_2_2_sidefx14AcesOutputFile11compressionEv
__ZNK14Imf_2_2_sidefx14RgbaOutputFile11compressionEv
__ZNK14Imf_2_2_sidefx18TiledRgbaInputFile11compressionEv
__ZNK14Imf_2_2_sidefx19TiledRgbaOutputFile11compressionEv
__ZNK14Imf_2_2_sidefx6Header11compressionEv
```

I suspect that this is something to do with IMF_EXPORT.

**Comment by Idclip** [comment]
Okay I know why this is happening. I'm using homebrew which puts everything under /include which is getting included before OpenEXR/IlmBase and ends up pulling in the non namespaced headers. I'll take a look at fixing the include order again, I must have broken it

**Comment by Idclip** [comment]
fixed on the stage branch
